From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Wed, 15 Nov 2023 12:57:36 -0700
Subject: [PATCH] Remove jackson

It's pointless to load two different json libraries when gson supports records now

diff --git a/build.gradle b/build.gradle
index b4304c7e505c88537aa3e204fb17c167f780af79..3ad891695265c560f7c6ffe56c6eca4002c4129d 100644
--- a/build.gradle
+++ b/build.gradle
@@ -120,7 +120,7 @@ dependencies {
 	// libraries
 	implementation libs.commons.io
 	implementation libs.gson
-	implementation libs.jackson
+	testImplementation libs.jackson
 	implementation libs.guava
 	implementation libs.bundles.asm
 
diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index c37ebd5c2ee12867a86e61bbc00aa21f20ab26a8..bd43efc8d35e8dadfbe06e5e81eca5a439ad2a60 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -27,8 +27,6 @@ package net.fabricmc.loom;
 import java.util.List;
 import java.util.Objects;
 
-import com.fasterxml.jackson.databind.DeserializationFeature;
-import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.collect.ImmutableMap;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
@@ -52,7 +50,6 @@ import net.fabricmc.loom.util.LibraryLocationLogger;
 
 public class LoomGradlePlugin implements BootstrappedPlugin {
 	public static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
-	public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
 	public static final String LOOM_VERSION = Objects.requireNonNullElse(LoomGradlePlugin.class.getPackage().getImplementationVersion(), "0.0.0+unknown");
 
 	/**
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/LayeredMappingsDependency.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/LayeredMappingsDependency.java
index 4ce75ef730ff3a5a98eb463af15bb646806ddd9a..7504e1b6e8d3721214aa5f79eae8a197cf7cc9c1 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/LayeredMappingsDependency.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/LayeredMappingsDependency.java
@@ -50,6 +50,7 @@ import net.fabricmc.loom.api.mappings.layered.MappingLayer;
 import net.fabricmc.loom.api.mappings.layered.MappingsNamespace;
 import net.fabricmc.loom.configuration.providers.mappings.extras.unpick.UnpickLayer;
 import net.fabricmc.loom.configuration.providers.mappings.utils.AddConstructorMappingVisitor;
+import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.ZipUtils;
 import net.fabricmc.mappingio.adapter.MappingDstNsReorder;
 import net.fabricmc.mappingio.adapter.MappingSourceNsSwitch;
@@ -118,7 +119,7 @@ public class LayeredMappingsDependency implements SelfResolvingDependency, FileC
 			return;
 		}
 
-		byte[] data = LoomGradlePlugin.OBJECT_MAPPER.writeValueAsString(signatureFixes).getBytes(StandardCharsets.UTF_8);
+		byte[] data = LoomGradlePlugin.GSON.toJson(signatureFixes, Constants.TypeTokens.MAP_STRING_STRING.getType()).getBytes(StandardCharsets.UTF_8);
 
 		ZipUtils.add(mappingsFile, "extras/record_signatures.json", data);
 	}
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
index 8321d3dec5880e13a3684939af689c55ee4220b0..462b4d65cfc01fe2b12b58ff08ea8ae25b69cafe 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/MappingConfiguration.java
@@ -234,8 +234,7 @@ public class MappingConfiguration {
 		}
 
 		try (Reader reader = Files.newBufferedReader(recordSignaturesJsonPath, StandardCharsets.UTF_8)) {
-			//noinspection unchecked
-			signatureFixes = LoomGradlePlugin.OBJECT_MAPPER.readValue(reader, Map.class);
+			signatureFixes = LoomGradlePlugin.GSON.fromJson(reader, Constants.TypeTokens.MAP_STRING_STRING.getType());
 		}
 	}
 
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/signatures/SignatureFixesLayerImpl.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/signatures/SignatureFixesLayerImpl.java
index 4328311b17f82f155158c39ef0026fd5f3a3b02a..bc9c449e2d032d94d5a698689162063a2bd933c7 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/signatures/SignatureFixesLayerImpl.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/signatures/SignatureFixesLayerImpl.java
@@ -31,6 +31,7 @@ import java.util.Map;
 import org.jetbrains.annotations.ApiStatus;
 
 import net.fabricmc.loom.api.mappings.layered.MappingLayer;
+import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.ZipUtils;
 import net.fabricmc.mappingio.MappingVisitor;
 
@@ -46,8 +47,7 @@ public record SignatureFixesLayerImpl(Path mappingsFile) implements MappingLayer
 	@Override
 	public Map<String, String> getSignatureFixes() {
 		try {
-			//noinspection unchecked
-			return ZipUtils.unpackJackson(mappingsFile(), SIGNATURE_FIXES_PATH, Map.class);
+			return ZipUtils.unpackJackson(mappingsFile(), SIGNATURE_FIXES_PATH, Constants.TypeTokens.MAP_STRING_STRING.getType());
 		} catch (IOException e) {
 			throw new RuntimeException("Failed to extract signature fixes", e);
 		}
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/unpick/UnpickLayer.java b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/unpick/UnpickLayer.java
index ce53e64bc4e6f73eedb811ef4d94e2b803a7cc32..960b32937c6d8ed8215605857f44cd6f8debbf6f 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/unpick/UnpickLayer.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/mappings/extras/unpick/UnpickLayer.java
@@ -46,7 +46,7 @@ public interface UnpickLayer {
 			final Metadata metadata;
 
 			try (Reader reader = Files.newBufferedReader(metadataPath, StandardCharsets.UTF_8)) {
-				metadata = LoomGradlePlugin.OBJECT_MAPPER.readValue(reader, Metadata.class);
+				metadata = LoomGradlePlugin.GSON.fromJson(reader, Metadata.class);
 			}
 
 			return new UnpickData(metadata, definitions);
@@ -54,7 +54,7 @@ public interface UnpickLayer {
 
 		public record Metadata(int version, String unpickGroup, String unpickVersion) {
 			public String asJson() throws IOException {
-				return LoomGradlePlugin.OBJECT_MAPPER.writeValueAsString(this);
+				return LoomGradlePlugin.GSON.toJson(this, Metadata.class);
 			}
 		}
 	}
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMetadataProvider.java b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMetadataProvider.java
index 396202cea72f59886a5a4228eeb6be3623bcc091..840207f620d1360dde4d34e02d998c1c48c32e17 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMetadataProvider.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftMetadataProvider.java
@@ -115,7 +115,7 @@ public final class MinecraftMetadataProvider {
 		}
 
 		final String versionManifest = builder.downloadString(cacheFile);
-		return LoomGradlePlugin.OBJECT_MAPPER.readValue(versionManifest, ManifestVersion.class);
+		return LoomGradlePlugin.GSON.fromJson(versionManifest, ManifestVersion.class);
 	}
 
 	private MinecraftVersionMeta readVersionMeta() throws IOException {
@@ -128,7 +128,7 @@ public final class MinecraftMetadataProvider {
 		}
 
 		final String json = builder.downloadString(options.minecraftMetadataPath());
-		return LoomGradlePlugin.OBJECT_MAPPER.readValue(json, MinecraftVersionMeta.class);
+		return LoomGradlePlugin.GSON.fromJson(json, MinecraftVersionMeta.class);
 	}
 
 	public record Options(String minecraftVersion,
diff --git a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/assets/AssetIndex.java b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/assets/AssetIndex.java
index 3dd6a3c5d52694625b95db4446f33fb0aa4af6f5..10f95c0dc2add0471c408e13c644e040dbd709f8 100644
--- a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/assets/AssetIndex.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/assets/AssetIndex.java
@@ -28,10 +28,8 @@ import java.util.Collection;
 import java.util.LinkedHashMap;
 import java.util.Map;
 
-import com.fasterxml.jackson.annotation.JsonProperty;
-
 @SuppressWarnings("unused")
-public record AssetIndex(Map<String, Entry> objects, boolean virtual, @JsonProperty("map_to_resources") boolean mapToResources) {
+public record AssetIndex(Map<String, Entry> objects, boolean virtual, boolean map_to_resources) {
 	public AssetIndex() {
 		this(new LinkedHashMap<>(), false, false);
 	}
diff --git a/src/main/java/net/fabricmc/loom/task/DownloadAssetsTask.java b/src/main/java/net/fabricmc/loom/task/DownloadAssetsTask.java
index b8abca181f087c23a26e41c82b7ffb135938b8f3..984be6ff80473eea3eea487044e4f898c2fac211 100644
--- a/src/main/java/net/fabricmc/loom/task/DownloadAssetsTask.java
+++ b/src/main/java/net/fabricmc/loom/task/DownloadAssetsTask.java
@@ -122,11 +122,11 @@ public abstract class DownloadAssetsTask extends AbstractLoomTask {
 				.sha1(assetIndex.sha1())
 				.downloadString(indexFile.toPath());
 
-		return LoomGradlePlugin.OBJECT_MAPPER.readValue(json, AssetIndex.class);
+		return LoomGradlePlugin.GSON.fromJson(json, AssetIndex.class);
 	}
 
 	private Path getAssetsPath(AssetIndex.Object object, AssetIndex index) {
-		if (index.mapToResources() || index.virtual()) {
+		if (index.map_to_resources() || index.virtual()) {
 			return new File(getLegacyResourcesDirectory().get().getAsFile(), object.path()).toPath();
 		}
 
diff --git a/src/main/java/net/fabricmc/loom/util/Constants.java b/src/main/java/net/fabricmc/loom/util/Constants.java
index a3590f0e8e2bf32fc85ee98d4969713c4076ad57..ff0220fa993979d6ff00370311bff00dafcd24e7 100644
--- a/src/main/java/net/fabricmc/loom/util/Constants.java
+++ b/src/main/java/net/fabricmc/loom/util/Constants.java
@@ -24,6 +24,9 @@
 
 package net.fabricmc.loom.util;
 
+import java.util.Map;
+
+import com.google.gson.reflect.TypeToken;
 import org.objectweb.asm.Opcodes;
 
 public class Constants {
@@ -125,4 +128,8 @@ public class Constants {
 		public static final String DISABLE_PROJECT_DEPENDENT_MODS = "fabric.loom.disableProjectDependentMods";
 		public static final String LIBRARY_PROCESSORS = "fabric.loom.libraryProcessors";
 	}
+
+	public static final class TypeTokens {
+		public static final TypeToken<Map<String, String>> MAP_STRING_STRING = new TypeToken<>() {};
+	}
 }
diff --git a/src/main/java/net/fabricmc/loom/util/LibraryLocationLogger.java b/src/main/java/net/fabricmc/loom/util/LibraryLocationLogger.java
index 2d27427f98152173af0e5862fb3cc4ffa3d516cb..ce193de42224f233803bcbfeb2934041c2cfd1c8 100644
--- a/src/main/java/net/fabricmc/loom/util/LibraryLocationLogger.java
+++ b/src/main/java/net/fabricmc/loom/util/LibraryLocationLogger.java
@@ -26,7 +26,6 @@ package net.fabricmc.loom.util;
 
 import java.util.List;
 
-import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.base.Preconditions;
 import com.google.gson.Gson;
 import kotlinx.metadata.jvm.KotlinClassMetadata;
@@ -53,7 +52,6 @@ public final class LibraryLocationLogger {
 			ClassRemapper.class,
 			ClassNode.class,
 			ASMifier.class,
-			ObjectMapper.class,
 			Gson.class,
 			Preconditions.class,
 			FileUtils.class
diff --git a/src/main/java/net/fabricmc/loom/util/ZipUtils.java b/src/main/java/net/fabricmc/loom/util/ZipUtils.java
index 6295d92313f0720ea278329b56029c04d6b8cb8c..15c59bbf3afa23fc68911bc7b306f5ca06b345b2 100644
--- a/src/main/java/net/fabricmc/loom/util/ZipUtils.java
+++ b/src/main/java/net/fabricmc/loom/util/ZipUtils.java
@@ -29,6 +29,7 @@ import java.io.DataInputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.io.UncheckedIOException;
+import java.lang.reflect.Type;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.LinkOption;
@@ -122,9 +123,9 @@ public class ZipUtils {
 		}
 	}
 
-	public static <T> T unpackJackson(Path zip, String path, Class<T> clazz) throws IOException {
+	public static <T> T unpackJackson(Path zip, String path, Type type) throws IOException {
 		final byte[] bytes = unpack(zip, path);
-		return LoomGradlePlugin.OBJECT_MAPPER.readValue(new String(bytes, StandardCharsets.UTF_8), clazz);
+		return LoomGradlePlugin.GSON.fromJson(new String(bytes, StandardCharsets.UTF_8), type);
 	}
 
 	public static void pack(Path from, Path zip) throws IOException {
